pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -195,7 +206,76 @@ func (a *genericActuator) Reconcile(ctx context.Context, worker *extensionsv1alp
 	return nil
 }
 
-func (a *genericActuator) deployMachineDeployments(ctx context.Context, cluster *controller.Cluster, worker *extensionsv1alpha1.Worker, existingMachineDeployments *machinev1alpha1.MachineDeploymentList, wantedMachineDeployments worker.MachineDeployments, classKind string, clusterAutoscalerUsed bool) error {
+func (a *genericActuator) addStateToMachineDeployment(ctx context.Context, worker *extensionsv1alpha1.Worker, wantedMachineDeployments workercontroller.MachineDeployments) error {
+	workerCopy := worker.DeepCopy()
+
+	if workerCopy.Status.State == nil || len(workerCopy.Status.State.Raw) <= 0 {
+		return nil
+	}
+	workerState := make(map[string]*workercontroller.MachineDeploymentState)
+
+	if err := json.Unmarshal(workerCopy.Status.State.Raw, &workerState); err != nil {
+		return err
+	}
+
+	for index, wantedMachineDeployment := range wantedMachineDeployments {
+		wantedMachineDeployments[index].State = workerState[wantedMachineDeployment.Name]
+	}
+	return nil
+}
+
+func (a *genericActuator) restore(ctx context.Context, cluster *controller.Cluster, worker *extensionsv1alpha1.Worker, existingMachineDeployments *machinev1alpha1.MachineDeploymentList, wantedMachineDeployments workercontroller.MachineDeployments) error {
+	existingMachineDeploymentNameSet := make(map[string]bool)
+
+	for _, existingMachineDeployment := range existingMachineDeployments.Items {
+		existingMachineDeploymentNameSet[existingMachineDeployment.Name] = true
+	}
+
+	for _, wantedMachineDeployment := range wantedMachineDeployments {
+
+		// We restore machineSet and machines only for missing machineDeployments
+		if existingMachineDeploymentNameSet[wantedMachineDeployment.Name] || wantedMachineDeployment.State == nil {
+			continue
+		}
+
+		if wantedMachineDeployment.State.MachineSet == nil {
+			continue
+		}
+
+		machineSet := &machinev1alpha1.MachineSet{}
+		if _, _, err := a.decoder.Decode(wantedMachineDeployment.State.MachineSet.Raw, nil, machineSet); err != nil {
+			return err
+		}
+
+		if err := a.client.Create(ctx, machineSet); err != nil && !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+
+		if err := a.client.Status().Update(ctx, machineSet); err != nil {
+			return err
+		}
+
+		for _, rawMachine := range wantedMachineDeployment.State.Machines {
+			if rawMachine.Raw == nil {
+				continue
+			}
+			machine := &machinev1alpha1.Machine{}
+			if _, _, err := a.decoder.Decode(rawMachine.Raw, nil, machine); err != nil {
+				return err
+			}
+			if err := a.client.Create(ctx, machine); err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
+			}
+			if err := a.client.Status().Update(ctx, machine); err != nil {
+				return err
+			}
+		}
+	}
+
```

Comment:
Don't we need to handle the owner reference UIDs here in the `restore` case? Or does MCM do this automatically itself correctly?

Reply:
At the moment MCM has a adoption mechanism and it assigns the right resource as owners.

Reply:
Great, thanks for clarifying and nice implementation @hardikdr @prashanth26 @ggaurav10